### PR TITLE
Singularity + particle bounding box stuff

### LINKED ISSUE #13062: Stage 2 singulo thinks it's bigger than it really is
Take a look at this code.

atom_2016-12-25_16-43-06

When the stage 2 singulo checks for shield generators to bounce off of, it checks 3 tiles away rather than 2. Essentially, singulo movement treats 3x3 singulos as if they were 5x5.

This results in two things:
One, like a really infuriating DVD screensaver, the stage 2 singulo will never bounce off the shield generators, always stop 1 tile before.
dreamseeker_2016-12-25_16-44-33

Two:
mpc-hc64_2016-12-25_16-46-35
mpc-hc64_2016-12-25_16-46-58
mpc-hc64_2016-12-25_16-47-14
If a singulo is created too close to the border, when it grows to stage 2, it will think it's 5x5, not 3x3 like it really is. So its movement checks the tiles on the OTHER SIDE of the containment field, and thus thinks it's OK to move that way.

Theoretically, fixing this would be as simple as changing this 3 to 2.
atom_2016-12-25_16-43-06

But it says "yes this is right", plus it's SINGULARITY CODE, maan, someone else do it

### PR DESCRIPTION
**Pull request by Sprok0**
The singularity is now actually the size it looks like, rather then being a 1x1 object in the center of a big sprite. Anything that moves onto the singularity is eaten immediately instead of waiting for the next process().
The singularity will now move away from containment fields before growing, instead of not growing at all if it was near a shield wall.
The particles launched by the PA are also now one 1x3 object instead of three 1x1 objects side by side. Particles are eaten by the first singularity they hit, or they'd add energy to them for each tile they move through.
Increased particle max range slightly.
Welder fuel tanks no longer explode when eaten by the singularity. The reason for this is as follows. When a singularity is hit by an ex_act(1), such as the middle of a welder fuel explosion, they have a 1/4 chance to be totally destroyed. This never happened in previous singulooses because the explosion needed to hit the actual singularity object in the center of the big sprite. Now any explosion happening on the singularity counts as hitting it, and each welder tank had a one in four chance of killing any singularity. If this is wanted I can change it back.

Fixes a runtime involving the multitile airlocks and shuttles that I though was related to the singularity but wasn't.
Fixes #13062

### CODE REVIEW

**Comment by ComicIronic** [comment]
Why not just change the whole block to do a divide and round up?

**Comment by 9600bauds** [comment]
maxresdefault

I'll test this in a while tho

**Comment by Sprok0** [comment]
The only real problem I see is that now that the 3x3 sing can hug the walls, if it shrinks to 1x1 on the far sides it'll be out of the PA's beam and unable to be easily restarted. I don't think it's really an issue, though.

File: code/__HELPERS/maths.dm [hunk 1]
```diff
@@ -241,3 +241,8 @@ proc/arctan(x)
 	var/mult = input/scale
 	var/trinum = (sqrt(8 * mult + 1) - 1 ) / 2
 	return trinum * scale
+
+/proc/round_up(num)
```
Comment (9600bauds) [inline code/__HELPERS/maths.dm]: Isn't this Ceiling()?

Comment (Sprok0) [reply code/__HELPERS/maths.dm]: Is it? I saw Ceiling but couldn't quickly make sense of it, and found this on the byond forum.

Comment (ComicIronic) [reply code/__HELPERS/maths.dm]: Yes, this is what ceiling should do.

**Comment by 9600bauds** [comment]
Well, it works as advertised. In hindsight since the singulo can hug the sides of the containment field this will also make it receive slightly less energy per second on average - unsure if that could make a stage 2 singulo revert to 1 with the PA set to 0? Will leave this running for a while and see if the singulo still exists when I come back.

Besides that, if there are any other problems this causes, I'm sure they won't go unnoticed for 5 years haha right??

**Comment by Probe1** [comment]
Sounds like its due for a christmas test merge.

**Comment by 9600bauds** [comment]
Haven't we had too many christmas test merges already

**Comment by Probe1** [comment]
That was more like a Christmas Hail Mary Full of Grace.

**Comment by 9600bauds** [comment]
Yeah this makes stage 2 singulos die with PA set to 0 unfortunately
dreamseeker_2016-12-25_21-00-34
probably the reason for the original shitty change in the first place

**Comment by Probe1** [comment]
Thank you for testing it thoroughly.

_cough_ _cough_

**Comment by 9600bauds** [comment]
To be fair Sprok0 thought of everything except that small detail more thoroughly than I did

Now the question is how to fix this

Lame-but-easiest-and-bestest solution is to just make it keep 1 tile of distance to the borders anyways and just fix the check

Other solutions would be increase the PA strength slightly

**Comment by Probe1** [comment]
The lamest fix is the best fix.

**Comment by 9600bauds** [comment]
but muh DVD screensaverrrr

**Comment by ComicIronic** [comment]
Why does a 3x3 die in that position? Surely it should still be eating PA waves?

**Comment by Sprok0** [comment]
The PA launches 3 particles side by side, at the very edges the singularity is only hit by one of them

**Comment by Intigracy** [comment]
You could make it fire 5 instead as a possible fix.

**Comment by Sprok0** [comment]
I'm looking into making it launch one three tile wide particle wave right now

File: code/__HELPERS/maths.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/__HELPERS/maths.dm]: Monsanto disapproves

**Comment by Sprok0** [comment]
This is becoming more involved then I thought it would. The problem was that only the actual singularity object in the center can be hit by particles, so the 3x3 isn't actually hit by anything if it's on the far edges. I'm trying to mess with bounding stuff so touching any of it works, but there are Issues and I might not be smart enough to do this.

**Comment by Sprok0** [comment]
The main problem right now is an offset between where it determines the center of the singularity is for purposes of consuming, and where the bounding box is. This only becomes desynced once the singularity moves once, if it stays still it works fine.
wrong

File: code/modules/power/singularity/singularity.dm [hunk 2]
```diff
@@ -91,6 +91,10 @@
 /obj/machinery/singularity/Bumped(atom/A)
 	consume(A)
 
+/obj/machinery/singularity/Crossed(atom/movable/A)
+	to_chat(world, "[src] crossed by \the [A]")
```
Comment (Iamgoofball) [inline code/modules/power/singularity/singularity.dm]: Debug

Comment (Sprok0) [reply code/modules/power/singularity/singularity.dm]: yes this isn't done yet

**Comment by 9600bauds** [comment]
Wow, that's fucked up
Good luck!

**Comment by Sprok0** [comment]
I've managed to fix the offset issue by replacing the singularity's normal movements with forceMove, can anyone see any downsides to that.

There are somewhat constant runtimes happening, which seem to be related to an object with non-default bound_x/y being moved. They happen on the main server with the 2x1 airlocks on the salvage shuttle whenever that moves.

**Comment by 9600bauds** [comment]
https://github.com/d3athrow/vgstation13/pull/13064#issuecomment-269138588

**Comment by Probe1** [comment]
boy I can't wait to merge this and see what happens.

I'll test it tomorrow.

**Comment by Sprok0** [comment]
don't merge this, a singulo just loosed randomly while testing and I'm trying to find out why

**Comment by Sprok0** [comment]
I've let one run for like an hour and it hasn't escaped, it seems safe now

**Pull request merged**